File: src/core/receipt.ts
/** Versioned proof receipt generation and verification. */

import { createHash, randomBytes } from 'node:crypto';
import { promises as fs } from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import { execSync } from 'node:child_process';
import { canonicalHash, type CanonicalizationProfile } from './canonical.js';
import { RECEIPT_PROTOCOL, validateReceiptEnvelope } from './envelope.js';
import { MerkleTree, type MerkleProfile } from './merkle.js';
import { ProofSigner, type SignatureAlgorithm, type SignatureRecord } from './signer.js';

export const RECEIPT_VERSION = '2.0.0' as const;
export const LEGACY_RECEIPT_VERSION = '1.0.0' as const;
export const RFC6962_MERKLE_ALGORITHM = 'RFC6962-SHA256' as const;
export const LEGACY_MERKLE_ALGORITHM = 'SHA-256' as const;

export interface ProofTask {
  name: string;
  description?: string;
  runner: string;
  status: 'SUCCESS' | 'FAILURE' | 'ATTESTED' | 'WARNING';
  exitCode: number;
  durationMs?: number;
  command?: string;
}

export interface ProofGitContext {
  commit: string;
  branch: string;
  dirty: boolean;
  remote?: string;
}

export interface ProofEnvironment {
  platform: string;
  arch: string;
  hostname?: string;
  runtime: string;
  git?: ProofGitContext;
}

export interface ProofArtifact {
  path: string;
  sha256: string;
  sizeBytes: number;
  mimeType?: string;
}

export interface ProofReceipt {
  version: typeof RECEIPT_VERSION | typeof LEGACY_RECEIPT_VERSION;
  protocol: typeof RECEIPT_PROTOCOL;
  proofId: string;
  timestamp: string;
  parentOrganization: string;
  task: ProofTask;
  environment: ProofEnvironment;
  artifacts: ProofArtifact[];
  merkle: {
    algorithm: typeof RFC6962_MERKLE_ALGORITHM | typeof LEGACY_MERKLE_ALGORITHM;
    leaves: string[];
    root: string;
  };
  signature: SignatureRecord;
  metadata: Record<string, unknown>;
}

export interface CreateReceiptOptions {
  task: Partial<ProofTask> & { name: string };
  artifacts?: Array<{ path: string; data?: Buffer | string; mimeType?: string }>;
  signingKey: string;
  signerIdentity: string;
  keyId?: string;
  algorithm?: SignatureAlgorithm;
  metadata?: Record<string, unknown>;
  cwd?: string;
  includeHostname?: boolean;
  /** Opt in only for a trusted working directory and Git installation. */
  includeGitContext?: boolean;
}

export interface VerificationResult {
  valid: boolean;
  trusted: boolean;
  merkleValid: boolean;
  signatureChecked: boolean;
  signatureValid: boolean | null;
  artifactsValid: boolean;
  errors: string[];
  warnings: string[];
  checkedArtifacts: number;
  receipt: ProofReceipt | null;
}

export type VerifyReceiptOptions = {
  checkFilesOnDisk?: boolean;
  cwd?: string;
} & (
  | { publicKeyOrSecret?: undefined; expectedAlgorithm?: undefined }
  | { publicKeyOrSecret: string; expectedAlgorithm: SignatureAlgorithm }
);

export function sanitizeGitRemote(remote: string): string | undefined {
  const trimmed = remote.trim();
  if (/^git@[A-Za-z0-9.-]+:[^\s?#]+$/.test(trimmed)) return trimmed;
  try {
    const parsed = new URL(trimmed);
    if (!['https:', 'http:', 'ssh:', 'git:'].includes(parsed.protocol)) return undefined;
    parsed.username = '';
    parsed.password = '';
    parsed.search = '';
    parsed.hash = '';
    return parsed.toString();
  } catch {
    return undefined;
  }
}

/** Gathers non-secret Git lineage without throwing outside a repository. */
export function getGitContext(cwd: string = process.cwd()): ProofGitContext | undefined {
  try {
    const commit = execSync('git rev-parse HEAD', { cwd, stdio: ['ignore', 'pipe', 'ignore'] })
      .toString().trim();
    const branch = execSync('git rev-parse --abbrev-ref HEAD', {
      cwd,
      stdio: ['ignore', 'pipe', 'ignore'],
    }).toString().trim();
    const dirty = execSync('git status --porcelain', {
      cwd,
      stdio: ['ignore', 'pipe', 'ignore'],
    }).toString().trim().length > 0;
    let remote: string | undefined;
    try {
      remote = sanitizeGitRemote(
        execSync('git config --get remote.origin.url', {
          cwd,
          stdio: ['ignore', 'pipe', 'ignore'],
        }).toString()
      );
    } catch {
      remote = undefined;
    }
    return { commit, branch, dirty, ...(remote === undefined ? {} : { remote }) };
  } catch {
    return undefined;
  }
}

export async function hashArtifact(
  artifactPath: string,
  data?: Buffer | string,
  cwd: string = process.cwd()
): Promise<ProofArtifact> {
  const resolved = path.isAbsolute(artifactPath) ? artifactPath : path.resolve(cwd, artifactPath);
  const buffer = data === undefined
    ? await fs.readFile(resolved)
    : typeof data === 'string'
      ? Buffer.from(data, 'utf8')
      : data;
  return {
    path: path.relative(cwd, resolved).replace(/\\/g, '/') || artifactPath,
    sha256: createHash('sha256').update(buffer).digest('hex'),
    sizeBytes: buffer.length,
  };
}

function profilesFor(version: ProofReceipt['version']): {
  canonical: CanonicalizationProfile;
  merkle: MerkleProfile;
} {
  return version === LEGACY_RECEIPT_VERSION
    ? { canonical: 'legacy', merkle: 'legacy-duplicated' }
    : { canonical: 'rfc8785', merkle: 'rfc6962' };
}

function receiptLeaves(
  task: ProofTask,
  environment: ProofEnvironment,
  artifacts: ProofArtifact[],
  version: ProofReceipt['version']
): string[] {
  const { canonical } = profilesFor(version);
  const artifactLeaves = version === LEGACY_RECEIPT_VERSION
    ? artifacts.map((artifact) => artifact.sha256)
    : artifacts.map((artifact) => canonicalHash(artifact, 'sha256', canonical));
  return [
    canonicalHash(task, 'sha256', canonical),
    canonicalHash(environment, 'sha256', canonical),
    ...artifactLeaves,
  ];
}

function signingPayload(receipt: Pick<
  ProofReceipt,
  'protocol' | 'version' | 'proofId' | 'timestamp' | 'parentOrganization' | 'task' | 'merkle' | 'metadata'
>, signature?: Pick<
  SignatureRecord,
  'algorithm' | 'keyId' | 'signerIdentity' | 'timestamp'
>): Record<string, unknown> {
  const legacyPayload = {
    proofId: receipt.proofId,
    timestamp: receipt.timestamp,
    parentOrganization: receipt.parentOrganization,
    taskName: receipt.task.name,
    merkleRoot: receipt.merkle.root,
  };
  if (receipt.version === LEGACY_RECEIPT_VERSION) return legacyPayload;
  if (signature === undefined) {
    throw new TypeError('Protocol v2 signing requires a complete signature header');
  }
  return {
    protocol: receipt.protocol,
    version: receipt.version,
    merkleAlgorithm: receipt.merkle.algorithm,
    metadataHash: canonicalHash(receipt.metadata),
    signatureAlgorithm: signature.algorithm,
    signatureKeyId: signature.keyId,
    signerIdentity: signature.signerIdentity,
    signatureTimestamp: signature.timestamp,
    ...legacyPayload,
  };
}

export async function createReceipt(options: CreateReceiptOptions): Promise<ProofReceipt> {
  const cwd = options.cwd ?? process.cwd();
  const proofId = `prf_${Date.now().toString(36)}_${randomBytes(6).toString('hex')}`;
  const timestamp = new Date().toISOString();
  const artifacts: ProofArtifact[] = [];
  for (const item of options.artifacts ?? []) {
    const artifact = await hashArtifact(item.path, item.data, cwd);
    artifacts.push(item.mimeType === undefined ? artifact : { ...artifact, mimeType: item.mimeType });
  }

  const git = options.includeGitContext === true ? getGitContext(cwd) : undefined;
  const environment: ProofEnvironment = {
    platform: os.platform(),
    arch: os.arch(),
    runtime: `node ${process.version}`,
    ...(options.includeHostname === true ? { hostname: os.hostname() } : {}),
    ...(git === undefined ? {} : { git }),
  };
  const task: ProofTask = {
    name: options.task.name,
    runner: options.task.runner ?? 'nymrel-agent',
    status: options.task.status ?? 'SUCCESS',
    exitCode: options.task.exitCode ?? 0,
    ...(options.task.description === undefined ? {} : { description: options.task.description }),
    ...(options.task.durationMs === undefined ? {} : { durationMs: options.task.durationMs }),
    ...(options.task.command === undefined ? {} : { command: options.task.command }),
  };

  const leaves = receiptLeaves(task, environment, artifacts, RECEIPT_VERSION);
  const merkleTree = new MerkleTree(leaves, { isPreHashed: true, profile: 'rfc6962' });
  const algorithm = options.algorithm ?? 'HMAC-SHA256';
  const keyId = options.keyId ?? (algorithm === 'Ed25519' ? 'ed25519-primary' : 'hmac-default');
  const metadata = options.metadata ?? {};
  const unsigned: Pick<
    ProofReceipt,
    'protocol' | 'version' | 'proofId' | 'timestamp' | 'parentOrganization' | 'task' | 'merkle' | 'metadata'
  > = {
    protocol: RECEIPT_PROTOCOL,
    version: RECEIPT_VERSION,
    proofId,
    timestamp,
    parentOrganization: 'Nymrel',
    task,
    merkle: {
      algorithm: RFC6962_MERKLE_ALGORITHM,
      leaves: merkleTree.getLeaves(),
      root: merkleTree.getRoot(),
    },
    metadata,
  };
  const signatureHeader = {
    algorithm,
    keyId,
    signerIdentity: options.signerIdentity,
    timestamp: new Date().toISOString(),
  };
  const signature: SignatureRecord = {
    ...signatureHeader,
    value: ProofSigner.signPayload(
      signingPayload(unsigned, signatureHeader),
      options.signingKey,
      algorithm,
      'rfc8785'
    ),
  };
  return {
    ...unsigned,
    environment,
    artifacts,
    signature,
    metadata,
  };
}

async function artifactPathWithin(cwd: string, artifactPath: string): Promise<string | undefined> {
  // Reject network/drive paths and lexical escapes before any filesystem lookup.
  if (path.posix.isAbsolute(artifactPath) || path.win32.isAbsolute(artifactPath) || /^[A-Za-z]:/.test(artifactPath)) return undefined;
  const unresolvedRoot = path.resolve(cwd);
  const portablePath = artifactPath.replace(/\\/g, '/');
  const lexicalRelative = path.relative(unresolvedRoot, path.resolve(unresolvedRoot, portablePath));
  if (lexicalRelative === '..' || lexicalRelative.startsWith(`..${path.sep}`) || path.isAbsolute(lexicalRelative)) return undefined;
  const root = await fs.realpath(unresolvedRoot).catch(() => unresolvedRoot);
  const unresolvedCandidate = path.resolve(root, portablePath);
  const candidate = await fs.realpath(unresolvedCandidate).catch(() => unresolvedCandidate);
  const relative = path.relative(root, candidate);
  if (relative === '' || (!relative.startsWith('..') && !path.isAbsolute(relative))) return candidate;
  return undefined;
}

export async function verifyReceipt(
  receiptInput: unknown,
  options: VerifyReceiptOptions = {}
): Promise<VerificationResult> {
  const invalidContext = (message: string): VerificationResult => ({
    valid: false, trusted: false, merkleValid: false, signatureChecked: false,
    signatureValid: null, artifactsValid: false, checkedArtifacts: 0, receipt: null,
    errors: [message], warnings: [],
  });
  if (!options || typeof options !== 'object' || Array.isArray(options)) {
    return invalidContext('Verification options must be an object');
  }
  const hasKey = options.publicKeyOrSecret != null;
  const hasAlgorithm = options.expectedAlgorithm != null;
  if (hasKey !== hasAlgorithm || (hasKey && (
    typeof options.publicKeyOrSecret !== 'string' || options.publicKeyOrSecret.length === 0 ||
    !['HMAC-SHA256', 'Ed25519'].includes(options.expectedAlgorithm as string)
  ))) {
    return invalidContext('A non-empty verification key and explicit expectedAlgorithm must be supplied together');
  }
  const envelope = validateReceiptEnvelope(receiptInput);
  if (!envelope.valid) {
    return {
      valid: false,
      trusted: false,
      merkleValid: false,
      signatureChecked: false,
      signatureValid: null,
      artifactsValid: false,
      errors: envelope.errors.map((error) =>
        `Envelope ${error.code}${error.path === null ? '' : ` at ${error.path}`}: ${error.message}`
      ),
      warnings: [],
      checkedArtifacts: 0,
      receipt: null,
    };
  }

  const receipt = receiptInput as ProofReceipt;
  if (hasKey && options.expectedAlgorithm !== receipt.signature.algorithm) {
    return invalidContext('Receipt signature algorithm does not match expectedAlgorithm');
  }
  const errors: string[] = [];
  const warnings: string[] = [];
  const cwd = options.cwd ?? process.cwd();
  const profiles = profilesFor(receipt.version);
  let expectedLeaves: string[];
  try {
    // Metadata is a signed v2 field even when this caller asks only for integrity.
    canonicalHash(receipt.metadata, 'sha256', profiles.canonical);
    expectedLeaves = receiptLeaves(receipt.task, receipt.environment, receipt.artifacts, receipt.version);
  } catch {
    return invalidContext('Receipt contains values outside its canonical JSON profile');
  }
  const calculatedTree = new MerkleTree(expectedLeaves, {
    isPreHashed: true,
    profile: profiles.merkle,
  });
  const calculatedRoot = calculatedTree.getRoot();
  const calculatedLeaves = calculatedTree.getLeaves();
  const leavesValid =
    calculatedLeaves.length === receipt.merkle.leaves.length &&
    calculatedLeaves.every(
      (leaf, index) => leaf === receipt.merkle.leaves[index].toLowerCase()
    );
  const merkleValid =
    leavesValid && calculatedRoot === receipt.merkle.root.toLowerCase();
  if (!leavesValid) errors.push('Merkle leaf list does not match the receipt payload');
  if (calculatedRoot !== receipt.merkle.root.toLowerCase()) {
    errors.push(
      `Merkle root mismatch: recorded '${receipt.merkle.root}', recalculated '${calculatedRoot}'`
    );
  }

  let artifactsValid = true;
  let checkedArtifacts = 0;
  if (options.checkFilesOnDisk === true) {
    for (const artifact of receipt.artifacts) {
      const fullPath = await artifactPathWithin(cwd, artifact.path);
      if (fullPath === undefined) {
        artifactsValid = false;
        errors.push(`Artifact path escapes verification root: '${artifact.path}'`);
        continue;
      }
      try {
        const fileData = await fs.readFile(fullPath);
        checkedArtifacts++;
        const actualHash = createHash('sha256').update(fileData).digest('hex');
        if (actualHash !== artifact.sha256.toLowerCase()) {
          artifactsValid = false;
          errors.push(`Artifact tampered: '${artifact.path}' (hash mismatch)`);
        }
      } catch (error: unknown) {
        artifactsValid = false;
        errors.push(
          `Artifact missing on disk: '${artifact.path}' (${error instanceof Error ? error.message : String(error)})`
        );
      }
    }
  }

  const signatureChecked = hasKey;
  let signatureValid: boolean | null = null;
  if (signatureChecked) {
    signatureValid = ProofSigner.verifySignature(
      signingPayload(receipt, receipt.signature),
      receipt.signature.value,
      options.publicKeyOrSecret as string,
      options.expectedAlgorithm as SignatureAlgorithm,
      profiles.canonical
    );
    if (!signatureValid) errors.push('Cryptographic signature verification failed with provided key');
  } else {
    warnings.push('Signature was not cryptographically verified (no public key or secret provided)');
  }

  if (receipt.version === LEGACY_RECEIPT_VERSION) {
    warnings.push('Legacy v1 authenticates artifact digests only, not artifact path/size/mimeType or artifact count, metadata, or signature identity fields');
  }

  const valid = errors.length === 0;
  return {
    valid,
    trusted: valid && signatureValid === true,
    merkleValid,
    signatureChecked,
    signatureValid,
    artifactsValid,
    errors,
    warnings,
    checkedArtifacts,
    receipt,
  };
}

File: python/nymrel_proof_ledger/receipt.py
"""Versioned proof receipt generation and verification."""

from __future__ import annotations

import hashlib
import os
import platform
import re
import subprocess
from datetime import UTC, datetime
from typing import Any
from urllib.parse import urlsplit, urlunsplit

from .canonical import canonical_hash
from .envelope import validate_receipt_envelope
from .merkle import MerkleTree
from .signer import ProofSigner

RECEIPT_PROTOCOL = "nymrel-proof-ledger"
RECEIPT_VERSION = "2.0.0"
LEGACY_RECEIPT_VERSION = "1.0.0"
RFC6962_MERKLE_ALGORITHM = "RFC6962-SHA256"
LEGACY_MERKLE_ALGORITHM = "SHA-256"


def sanitize_git_remote(remote: str) -> str | None:
    trimmed = remote.strip()
    if re.fullmatch(r"git@[A-Za-z0-9.-]+:[^\s?#]+", trimmed):
        return trimmed
    try:
        parsed = urlsplit(trimmed)
        if parsed.scheme not in ("https", "http", "ssh", "git") or not parsed.hostname:
            return None
        host = parsed.hostname
        if parsed.port is not None:
            host = f"{host}:{parsed.port}"
        return urlunsplit((parsed.scheme, host, parsed.path, "", ""))
    except (ValueError, TypeError):
        return None


def get_git_context(cwd: str | None = None) -> dict[str, Any] | None:
    cwd = cwd or os.getcwd()
    try:
        commit = (
            subprocess.check_output(
                ["git", "rev-parse", "HEAD"], cwd=cwd, stderr=subprocess.DEVNULL
            )
            .decode("utf-8")
            .strip()
        )
        branch = (
            subprocess.check_output(
                ["git", "rev-parse", "--abbrev-ref", "HEAD"],
                cwd=cwd,
                stderr=subprocess.DEVNULL,
            )
            .decode("utf-8")
            .strip()
        )
        dirty = bool(
            subprocess.check_output(
                ["git", "status", "--porcelain"], cwd=cwd, stderr=subprocess.DEVNULL
            )
            .decode("utf-8")
            .strip()
        )
        remote = None
        try:
            remote = sanitize_git_remote(
                subprocess.check_output(
                    ["git", "config", "--get", "remote.origin.url"],
                    cwd=cwd,
                    stderr=subprocess.DEVNULL,
                ).decode("utf-8")
            )
        except (OSError, subprocess.SubprocessError):
            remote = None
        return {
            "commit": commit,
            "branch": branch,
            "dirty": dirty,
            **({} if remote is None else {"remote": remote}),
        }
    except (OSError, subprocess.SubprocessError):
        return None


def hash_artifact(
    artifact_path: str,
    data: str | bytes | None = None,
    cwd: str | None = None,
) -> dict[str, Any]:
    cwd = cwd or os.getcwd()
    resolved = (
        artifact_path
        if os.path.isabs(artifact_path)
        else os.path.abspath(os.path.join(cwd, artifact_path))
    )
    if data is None:
        with open(resolved, "rb") as file:
            buffer = file.read()
    else:
        buffer = data.encode("utf-8") if isinstance(data, str) else bytes(data)
    try:
        relative = os.path.relpath(resolved, cwd).replace("\\", "/")
    except ValueError:
        relative = artifact_path
    return {
        "path": relative or artifact_path,
        "sha256": hashlib.sha256(buffer).hexdigest(),
        "sizeBytes": len(buffer),
    }


def _profiles(version: str) -> tuple[str, str]:
    return (
        ("legacy", "legacy-duplicated")
        if version == LEGACY_RECEIPT_VERSION
        else ("rfc8785", "rfc6962")
    )


def _receipt_leaves(
    task: dict[str, Any],
    environment: dict[str, Any],
    artifacts: list[dict[str, Any]],
    version: str,
) -> list[str]:
    canonical_profile, _ = _profiles(version)
    artifact_leaves = (
        [artifact["sha256"] for artifact in artifacts]
        if version == LEGACY_RECEIPT_VERSION
        else [
            canonical_hash(artifact, profile=canonical_profile)
            for artifact in artifacts
        ]
    )
    return [
        canonical_hash(task, profile=canonical_profile),
        canonical_hash(environment, profile=canonical_profile),
        *artifact_leaves,
    ]


def _signing_payload(
    receipt: dict[str, Any],
    signature: dict[str, Any] | None = None,
) -> dict[str, Any]:
    legacy_payload = {
        "proofId": receipt["proofId"],
        "timestamp": receipt["timestamp"],
        "parentOrganization": receipt["parentOrganization"],
        "taskName": receipt["task"]["name"],
        "merkleRoot": receipt["merkle"]["root"],
    }
    if receipt["version"] == LEGACY_RECEIPT_VERSION:
        return legacy_payload
    if signature is None:
        raise ValueError("Protocol v2 signing requires a complete signature header")
    return {
        "protocol": receipt["protocol"],
        "version": receipt["version"],
        "merkleAlgorithm": receipt["merkle"]["algorithm"],
        "metadataHash": canonical_hash(receipt["metadata"]),
        "signatureAlgorithm": signature["algorithm"],
        "signatureKeyId": signature["keyId"],
        "signerIdentity": signature["signerIdentity"],
        "signatureTimestamp": signature["timestamp"],
        **legacy_payload,
    }


def create_receipt(
    task: dict[str, Any],
    signing_key: str,
    signer_identity: str,
    artifacts: list[dict[str, Any]] | None = None,
    key_id: str | None = None,
    algorithm: str = "HMAC-SHA256",
    metadata: dict[str, Any] | None = None,
    cwd: str | None = None,
    include_hostname: bool = False,
) -> dict[str, Any]:
    cwd = cwd or os.getcwd()
    now = datetime.now(UTC)
    proof_id = f"prf_{int(now.timestamp()):x}_{os.urandom(6).hex()}"
    timestamp = now.isoformat().replace("+00:00", "Z")
    processed_artifacts: list[dict[str, Any]] = []
    for item in artifacts or []:
        artifact = hash_artifact(item["path"], item.get("data"), cwd)
        if "mimeType" in item:
            artifact["mimeType"] = item["mimeType"]
        processed_artifacts.append(artifact)

    environment: dict[str, Any] = {
        "platform": platform.system().lower(),
        "arch": platform.machine().lower(),
        "runtime": f"python {platform.python_version()}",
    }
    if include_hostname:
        environment["hostname"] = platform.node()
    git = get_git_context(cwd)
    if git is not None:
        environment["git"] = git

    task_record: dict[str, Any] = {
        "name": task["name"],
        "runner": task.get("runner", "nymrel-agent"),
        "status": task.get("status", "SUCCESS"),
        "exitCode": task.get("exitCode", 0),
    }
    for optional in ("description", "durationMs", "command"):
        if optional in task:
            task_record[optional] = task[optional]

    leaves = _receipt_leaves(
        task_record, environment, processed_artifacts, RECEIPT_VERSION
    )
    tree = MerkleTree(leaves, is_pre_hashed=True, profile="rfc6962")
    receipt_metadata = metadata or {}
    unsigned: dict[str, Any] = {
        "protocol": RECEIPT_PROTOCOL,
        "version": RECEIPT_VERSION,
        "proofId": proof_id,
        "timestamp": timestamp,
        "parentOrganization": "Nymrel",
        "task": task_record,
        "merkle": {
            "algorithm": RFC6962_MERKLE_ALGORITHM,
            "leaves": tree.get_leaves(),
            "root": tree.get_root(),
        },
        "metadata": receipt_metadata,
    }
    resolved_key_id = key_id or (
        "ed25519-primary" if algorithm == "Ed25519" else "hmac-default"
    )
    signature_header = {
        "algorithm": algorithm,
        "keyId": resolved_key_id,
        "signerIdentity": signer_identity,
        "timestamp": datetime.now(UTC).isoformat().replace("+00:00", "Z"),
    }
    signature = {
        **signature_header,
        "value": ProofSigner.sign_payload(
            _signing_payload(unsigned, signature_header),
            signing_key,
            algorithm,
            "rfc8785",
        ),
    }
    return {
        **unsigned,
        "environment": environment,
        "artifacts": processed_artifacts,
        "signature": signature,
        "metadata": receipt_metadata,
    }


def _artifact_path_within(cwd: str, artifact_path: str) -> str | None:
    root = os.path.realpath(os.path.abspath(cwd))
    candidate = os.path.realpath(os.path.abspath(os.path.join(root, artifact_path)))
    try:
        return candidate if os.path.commonpath((root, candidate)) == root else None
    except ValueError:
        return None


def verify_receipt(
    receipt: Any,
    public_key_or_secret: str | None = None,
    check_files_on_disk: bool = False,
    cwd: str | None = None,
) -> dict[str, Any]:
    envelope = validate_receipt_envelope(receipt)
    if not envelope["valid"]:
        return {
            "valid": False,
            "trusted": False,
            "merkleValid": False,
            "signatureChecked": False,
            "signatureValid": None,
            "artifactsValid": False,
            "errors": [
                f"Envelope {error['code']}"
                + ("" if error["path"] is None else f" at {error['path']}")
                + f": {error['message']}"
                for error in envelope["errors"]
            ],
            "warnings": [],
            "checkedArtifacts": 0,
            "receipt": None,
        }

    errors: list[str] = []
    warnings: list[str] = []
    cwd = cwd or os.getcwd()
    canonical_profile, merkle_profile = _profiles(receipt["version"])
    expected_raw_leaves = _receipt_leaves(
        receipt["task"],
        receipt["environment"],
        receipt["artifacts"],
        receipt["version"],
    )
    calculated_tree = MerkleTree(
        expected_raw_leaves, is_pre_hashed=True, profile=merkle_profile
    )
    calculated_leaves = calculated_tree.get_leaves()
    recorded_leaves = [leaf.lower() for leaf in receipt["merkle"]["leaves"]]
    leaves_valid = calculated_leaves == recorded_leaves
    calculated_root = calculated_tree.get_root()
    recorded_root = receipt["merkle"]["root"].lower()
    merkle_valid = leaves_valid and calculated_root == recorded_root
    if not leaves_valid:
        errors.append("Merkle leaf list does not match the receipt payload")
    if calculated_root != recorded_root:
        errors.append(
            f"Merkle root mismatch: recorded '{receipt['merkle']['root']}', "
            f"recalculated '{calculated_root}'"
        )

    artifacts_valid = True
    checked_artifacts = 0
    if check_files_on_disk:
        for artifact in receipt["artifacts"]:
            full_path = _artifact_path_within(cwd, artifact["path"])
            if full_path is None:
                artifacts_valid = False
                errors.append(
                    f"Artifact path escapes verification root: '{artifact['path']}'"
                )
                continue
            try:
                with open(full_path, "rb") as file:
                    actual_hash = hashlib.sha256(file.read()).hexdigest()
                checked_artifacts += 1
                if actual_hash != artifact["sha256"].lower():
                    artifacts_valid = False
                    errors.append(
                        f"Artifact tampered: '{artifact['path']}' (hash mismatch)"
                    )
            except OSError as error:
                artifacts_valid = False
                errors.append(
                    f"Artifact missing on disk: '{artifact['path']}' ({error})"
                )

    signature_checked = public_key_or_secret is not None
    signature_valid: bool | None = None
    if signature_checked:
        signature_valid = ProofSigner.verify_signature(
            _signing_payload(receipt, receipt["signature"]),
            receipt["signature"]["value"],
            public_key_or_secret,
            receipt["signature"]["algorithm"],
            canonical_profile,
        )
        if not signature_valid:
            errors.append(
                "Cryptographic signature verification failed with provided key"
            )
    else:
        warnings.append(
            "Signature was not cryptographically verified (no public key or secret provided)"
        )

    valid = not errors
    return {
        "valid": valid,
        "trusted": valid and signature_valid is True,
        "merkleValid": merkle_valid,
        "signatureChecked": signature_checked,
        "signatureValid": signature_valid,
        "artifactsValid": artifacts_valid,
        "errors": errors,
        "warnings": warnings,
        "checkedArtifacts": checked_artifacts,
        "receipt": receipt,
    }
